Framework/DataHandling/src/LoadDNSEvent.cpp

```diff
@@ -0,0 +1,479 @@
+
+  FileByteStream file(static_cast<std::string>(fileName));
+
+  auto finalEventAccumulator = parse_File(file, fileName);
+  populate_EventWorkspace(outputWS, finalEventAccumulator);
+  if (m_setBinBoundary && (outputWS->getNumberEvents() > 0)) {
+    outputWS->setAllX({0, outputWS->getEventXMax()});
+  }
+  setProperty("OutputWorkspace", outputWS);
+}
+
+void LoadDNSEvent::populate_EventWorkspace(EventWorkspace_sptr &eventWS, EventAccumulator &finalEventAccumulator) {
+  static const unsigned EVENTS_PER_PROGRESS = 100;
+  // The number of steps depends on the type of input file
+  Progress progress(this, 0.0, 1.0, finalEventAccumulator.neutronEvents.size() / EVENTS_PER_PROGRESS);
+
+  // Sort reversed (latest event first, most early event last):
+  std::sort(finalEventAccumulator.triggerEvents.begin(), finalEventAccumulator.triggerEvents.end(),
+            [](auto l, auto r) { return l.timestamp > r.timestamp; });
+
+  std::atomic<unsigned int> oversizedChanelIndexCounterA(0);
+  std::atomic<unsigned int> oversizedPosCounterA(0);
+
+  PARALLEL_FOR_IF(Kernel::threadSafe(*eventWS))
+  for (int eventIndex = 0; eventIndex < static_cast<int>(finalEventAccumulator.neutronEvents.size()); ++eventIndex) {
+    // uint64_t chopperTimestamp = 0;
+    unsigned int oversizedChanelIndexCounter = 0;
+    unsigned int oversizedPosCounter = 0;
+    const auto wsIndex = eventIndex;
+    auto &eventList = finalEventAccumulator.neutronEvents[eventIndex];
+    if (eventList.size() != 0) {
+      std::sort(eventList.begin(), eventList.end(), [](auto l, auto r) { return l.timestamp < r.timestamp; });
+    }
+
+    auto chopperIt = finalEventAccumulator.triggerEvents.cbegin();
+
+    auto &spectrum = eventWS->getSpectrum(wsIndex);
+    PARALLEL_START_INTERUPT_REGION
+
+    uint64_t numProcessed = 0;
+    for (const auto &event : eventList) {
+      numProcessed++;
+      if (numProcessed % EVENTS_PER_PROGRESS == 0) {
+        progress.report();
+        if (this->getCancel()) {
+          throw CancelException();
+        }
+      }
+
+      chopperIt =
+          std::lower_bound(finalEventAccumulator.triggerEvents.cbegin(), finalEventAccumulator.triggerEvents.cend(),
+                           event.timestamp, [](auto l, auto r) { return l.timestamp > r; });
+      const uint64_t chopperTimestamp = chopperIt != finalEventAccumulator.triggerEvents.cend()
+                                            ? chopperIt->timestamp
+                                            : 0; // before first chopper trigger
+      if ((chopperTimestamp == 0) && m_discardPreChopperEvents) {
+        // throw away events before first chopper trigger
+        continue;
+      }
+      spectrum.addEventQuickly(Types::Event::TofEvent(double(event.timestamp - chopperTimestamp) / 10.0));
+    }
+
+    PARALLEL_END_INTERUPT_REGION
+    oversizedChanelIndexCounterA += oversizedChanelIndexCounter;
+    oversizedPosCounterA += oversizedPosCounter;
+    PARALLEL_CHECK_INTERUPT_REGION
```

Comment:
`CHECK_INTERUPT_REGION` neds to move after the loop (swap it with the line below)

A nice example I refer back to often can be found here: https://developer.mantidproject.org/MultiThreadingInAlgorithms.html

Reply:
ah I understand